File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

long deleteLastNumber(long);
int sumOfDigits(int);
void printCardType(long);
int getFirst2DigitNumber(long);
void printCardType(long);

int counterDigit = 0;

int main(void)
{
    long cardNumber = get_long("Number: ");
    
    long tmpNumber = cardNumber;
    int sum = 0, sum1 = 0, sum2 = 0;
    
    while(tmpNumber != 0)
    {
        counterDigit++;
        if (counterDigit % 2 == 1)
        {
            sum1 += tmpNumber % 10;
            tmpNumber = deleteLastNumber(tmpNumber);
            
        }
        else
        {
            sum2 += sumOfDigits((tmpNumber % 10) * 2);
            tmpNumber = deleteLastNumber(tmpNumber);
        }
    }
    sum = sum1 + sum2;
    if (counterDigit < 15 || counterDigit > 16)
        printf("INVALID\n");
    else if (sum % 10 == 0)
        printCardType(cardNumber);
    else
        printf("INVALID\n");
       

}

int getFirst2DigitNumber(long cardNumber)
{
    for(int i = 0; i < counterDigit - 2; i++)
        cardNumber = deleteLastNumber(cardNumber);
    return cardNumber;
}

void printCardType(long number)
{
    int control = getFirst2DigitNumber(number);
    
    if (control == 34 || control == 37)
        printf("AMEX\n");
    else if (control == 51 || control == 52 || control == 53 || control == 54 || control == 55)
        printf("MASTERCARD\n");
    else if (control / 10 == 4)
        printf("VISA\n");
    else
        printf("INVALID\n");
}
int sumOfDigits(int number){
    int sum = 0;
    while(number != 0)
    {
        sum += number % 10;
        number /= 10;
    }
    return sum;
}

long deleteLastNumber(long number)
{
    number /= 10;
    return number;
}

File: pset6/credit/credit.py
from cs50 import get_int


def main():
    n = 0
    while True:
        n = get_int("Number: ")
        if n > 0:
            break
    print(get_card_type(n))


def get_card_type(n):
    if not is_real_credit(str(n)):
        return "INVALID"
    n = str(n)
    n = int(n[:2])  # ilk iki haneyi al
    if n == 34 or n == 37:
        return "AMEX"
    elif n >= 51 and n <= 55:
        return "MASTERCARD"
    elif n // 10 == 4:
        return "VISA"
    else:
        return "INVALID"


def sum_of_digits(n):
    sum = 0
    for digit in str(n):
        sum += int(digit)
    return sum


def is_real_credit(str):
    r_str = str[::-1]  # sayıyı ters çevir
    i = 0
    sum = 0
    for c in r_str:
        if i % 2 == 1:
            sum += sum_of_digits(int(c) * 2)
        i += 1
    i = 0
    for c in r_str:
        if i % 2 == 0:
            sum += int(c)
        i += 1
    if sum % 10 == 0 and len(str) >= 13:
        return True
    else:
        return False


if __name__ == "__main__":
    main()